gem/gem.py

```diff
@@ -891,15 +935,40 @@ def __new__(cls, array):
         dtype = Node.inherit_dtype_from_children(tuple(array.flat))
 
         # Handle children with shape
-        child_shape = array.flat[0].shape
+        e0 = array.flat[0]
+        child_shape = e0.shape
         assert all(elem.shape == child_shape for elem in array.flat)
 
+        # Simplify [v[multiindex, j] for j in range(n)] -> partial_indexed(v, multiindex)
+        if all(isinstance(elem, Indexed) for elem in array.flat):
+            tensor = e0.children[0]
+            multiindex = tuple(i for i in e0.multiindex if not isinstance(i, Integral))
+            index_shape = tuple(i.extent for i in multiindex if isinstance(i, Index))
+            if index_shape + array.shape + child_shape == tensor.shape:
+                if all(elem.children[0] == tensor for elem in array.flat[1:]):
+                    if all(elem.multiindex == multiindex + idx for idx, elem in numpy.ndenumerate(array)):
+                        return partial_indexed(tensor, multiindex)
```

Comment:
Maybe we should make it more explicit that we are only handling multiindex of the following pattern: `(Index, Index, Index, ..., Integral, Integral, ...)`. It looks the case where we have `VariableIndex`s, for instance, is handled in a very obscure way.